src/shogun/clustering/KMeansBase.cpp

```diff
@@ -130,6 +130,44 @@ void KMeansBase::compute_cluster_variances()
 	}
 }
 
+SGMatrix<float64_t> KMeansBase::compute_std_dev() const
+{
+	require(cluster_centers.size() > 0, "KMeans is not trained!");
+
+	SGMatrix<float64_t> points = distance->get_rhs()
+	                                 ->as<DenseFeatures<float64_t>>()
+	                                 ->get_feature_matrix();
+	SGVector<float64_t> cluster_assignments = const_cast<KMeansBase*>(this)
```

Comment:
please no `const_cast`. @karlnapf this is probably more of an indication that we need to redesign KMeans no?

Reply:
probably, but can't we avoid the cast otherwise?

Reply:
I think we have to either make this function non const or make apply const

Reply:
Making it nonconst breaks watch_method. I tried to make _apply_ const but it causes continuous changes in the derived classes of DistanceMachine. I was not sure that these changes will not bring something bad and decided to use the most obvious solution.

Reply:
What do you mean it breaks? I think it should work...

Reply:
If no _const_ keyword is present:
```
terminate called after throwing an instance of 'shogun::ShogunException'
  what():  The parameter KMeans::std_dev is registered as a function, use the .run() method instead!
```

Reply:
apply should be const, BUT we cannot do that easily. So need to find another solution. Const cast is not the best. Maybe gil has an idea re the watch stuff

Reply:
Could you maybe write a const helper method that computes the assignments, and then you call that from `apply` and from this place here?

Reply:
The error message tells you what to do though... it’s not breaking. Instead of get you should use run

Reply:
run is void, no?

Reply:
I don’t think so. Check in SGObject

Reply:
Hmm, you’re right. Not sure then :D

Reply:
`void run(std::string_view name) const noexcept(false)`
Am I miss something?

Reply:
no, I just got confused, sorry! what @karlnapf suggested should work though! :)